### DISCUSSION TITLE
Idea: Check allocation code during compilation

### DISCUSSION THREAD

**Discussion by Romfos** [dotnet/csharplang#6202] [Language Ideas]
Hi,

Could you please implement simple compile-time check for code that do memory allocation?

Basic idea:
1) [AllocationNonFree] - attribute for method\propeties\any other type member that compiler will be able to recognize as API that do memory allocation
2) [AllocationFree] - attribute method\propeties\any other type member that compiler will be able to recognize as API that don't do memory allocation. If method with this attribute have calls to other methods\properties that have attribute [AllocationNonFree] - show warning in this location

This attributes and diagnostic will help us to make code base better and little bit easier to develop\extend\support :) 

Thank you

Similar idea could with speed-optimized code. For example some functions in .NET have 'classic' and more performant alternative. Idea is to make attribute that will be suggest more performase-optimal API for code with [PerformanceOptimized] (for example) attribute on method

cc @davidfowl @MadsTorgersen

**Comment by CyrusNajmabadi** [comment]
This would be something that could likely be done with an analyzer/attribute combo.  it's probably not ideal for adding to the language itself.  :)

**Comment by Romfos** [comment]
This feature will be more usefull if BCL will be annotated like this and automatically mark other method that have dependencies on it

**Reply by CyrusNajmabadi** [reply]
@Romfos BCL suggestions can be made at: github.com/dotnet/runtime.  This is the repo for making language changes to C# :)

**Reply by Romfos** [reply]
https://github.com/dotnet/runtime/issues/70387
